examples/analytics-3p-remote-frame-helper.js

```diff
@@ -0,0 +1,188 @@
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+const RTV_VERSION = '1234'; // TODO Replace with import from AMP, or just remove
+
+/**
+ * This provides the "glue" between the AMP Analytics tag and the third-party
+ * vendor's metrics-collection page.
+ *
+ * The vendor's page must implement onNewAmpAnalyticsInstance().
+ */
+class AmpAnalyticsRemoteFrameManager {
+  constructor() {
+    /**
+     * @type {Function<!String>}
+     * @private
+     */
+    this.listener_ = () => {};
+
+    /**
+     * @type {Function<!String>}
+     * @private
+     */
+    this.extraDataListener_ = () => {};
+
+    /**
+     * Holds a mapping between sender ID and extra data
+     * @type {Map}
+     * @private
+     */
+    this.senderIdToExtraData_ = {}; // TODO: Replace with map from AMP
+  }
+
+  /**
+   * Registers a callback function to be called when AMP Analytics events occur
+   * @param {!Function} listener A function that takes an array of event
+   * strings, and does something with them.
+   * @param {?Function} extraDataListener A function to receive any extra
+   * setup data sent by the creative
+   */
+  registerAmpAnalyticsEventListener(listener, extraDataListener) {
+    this.listener_ = listener;
+    this.extraDataListener_ = extraDataListener;
+  }
+
+  /**
+   * Receives a message from the creative. Handles deserialization (which
+   * filters out messages not actually meant for us) and passes it to the
+   * listener that was supplied to registerAmpAnalyticsEventListener above.
+   * @param {!String} message The message that was received.
+   */
+  receiveMessageFromCreative(message) {
+    if (message && message.data) {
+      const deserialized = deserializeMessage(message.data);
+      if (deserialized) {
+        this.setSentinel_(deserialized.sentinel);
+        if (deserialized.ampAnalyticsExtraData) {
+          this.senderIdToExtraData_[deserialized.senderId] =
+            deserialized.ampAnalyticsExtraData;
+          window.requestIdleCallback(() => {
+            this.extraDataListener_(deserialized.senderId,
+              deserialized.ampAnalyticsExtraData);
+          });
+        } else if (deserialized.ampAnalyticsEvents) {
+          window.requestIdleCallback(() => {
+            this.listener_(deserialized.ampAnalyticsEvents);
+          });
+        }
+      }
+    }
+  }
+
+  /**
+   * Sets the sentinel value, which is used to identify which creative the
+   * event messages come from
+   * @param {!String} sentinel The sentinel value
+   */
+  setSentinel_(sentinel) {
+    if (this.sentinel_ && sentinel != this.sentinel_) {
+      console.warn("Attempting to set sentinel to " + sentinel + " when it" +
+          " was already set to " + this.sentinel_);
+    }
+    this.sentinel_ = sentinel;
```

Comment:
The sentinel is created in transport.js, in createCrossDomainIframe():
iframeMessagingClient.setSentinel(Transport.createUniqueId_());
So each frame (since we may have multiple vendors) gets a unique sentinel, which associated to the frame URL in a map that is static to the Transport class. And then this is indeed passed to the 3p frame (this file is running in the 3p frame).